Doc/howto/concurrency.rst

```diff
@@ -0,0 +1,1451 @@
+
+.. _python-interpreters-overhead:
+
+Improving Performance for Multiple Interpreters
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The long effort to improve on Python's implementation of multiple
+interpreters focused on isolation and stability.  There was very little
+done to improve performance.  This has the most impact on:
+
+* how much memory each interpreter uses
+  (i.e. how many can run at the same time)
+* how long it takes to create a new interpreter
+
+As the work on isolation wraps up, improvements will shift to focus
+on performance and memory usage.  Thus the overhead associated with
+using multiple interpreters will drastically decrease over time.
+
+.. _python-coroutines:
+
+Coroutines (Async/Await)
+------------------------
+
+.. currentmodule:: asyncio
+
+The use of :term:`coroutines <coroutine>` for concurrency has been
+around a long time and has grown in popularity in the software world,
+particularly with the addition of ``async/await`` syntax in
+various languages.
+
+Python has supported coroutines to some degree since the beginning.
+The best example is :pypi:`twisted`, which has provided this concurrency
+model for decades.  For most of that time :pypi:`!twisted` did it
+primarily through callbacks and a form of "promises"/"futures".
+
+Explicit support for coroutines in Python really started with the
+introduction of :term:`generators <generator>` in Python 2.2
+(:pep:`255`).  In Python 2.5 (:pep:`342`), :term:`!generators` were
+tweaked to explicitly support use as coroutines.  That went a step
+further in Python 3.3 with the addition of ``yield from`` (:pep:`380`)
+and the :mod:`asyncio` module (:pep:`3156`).  Finally, in Python 3.5
+(:pep:`492`), we got dedicated ``async/await`` syntax
+and :ref:`a dedicated protocol <coroutine-protocol>`
+for :term:`!coroutine` objects.
+
+There are three main pieces to using coroutines:
+
+* coroutines (non-blocking, yield control instead)
+* an event loop (schedules coroutines)
+* coroutine wrappers around blocking operations
+
+A :term:`coroutine function` looks *almost* the same as a regular
+function.  It is a non-blocking function that *cooperatively* yields
+control of the program to other coroutines, which in turn yield control
+back (eventually).  At those points of synchronization,
+coroutines often provide data to one another.
+
+The event loop is what keeps track of which coroutines have yielded
+control and which should get control next.
+
+Generally a coroutine needs to avoid doing anything that takes very long
+before yielding control back to the event loop.  Any blocking operation
+in a coroutine, like waiting on a socket, has to be implemented in a way
+that only waits a little while, yields, and then waits again, etc. until
+ready.  The alternative is to wrap the blocking operation/function
+in some sort of "future" coroutine that yields until the blocking
+operation completes.  The event loop can also fill that role
+to an extent.
```

Comment:
```suggestion
Generally, a coroutine needs to avoid doing anything that takes a long time
before yielding control back to the event loop.  Any blocking operation
in a coroutine, like waiting on a socket, has to be implemented in a way
that only waits a little while, yields, and then waits again, etc., until
ready.  The alternative is to wrap the blocking operation/function
in a "future" coroutine that yields until the blocking
operation completes.  The event loop can also fill that role
to an extent.
```

Reply:
I'm dropping this.